proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,532 @@
+
+assert(r.p1 instanceof Point);
+assert(r.p2 instanceof Point);
+```
+
+Without knowing anything about `wasmInstance`, can JS assume *based solely on
+the field types of `Rect`* that the asserts hold?
+
+If the JS `ref` Value Types were unified with the WebAssembly
+[`ref` type constructor](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types),
+the answer would be "no": `wasmInstance` could assign any typed object with two
+`i32` fields, including those with a `null` `__proto__` and thus no `x` and `y`
+property names. This would be surprising from a JS perspective and would also
+give up future WebAssembly Host Binding optimizations that are mentioned
+at the end of this section.
+
+To avoid problem, as a starting point, the following WebAssembly module:
+
+```wat
+;; example3.wat --> example3.wasm
+(module
+    (type $P (struct (field i32) (field i32)))
+    (type $R (struct (field $x1 (ref $P)) (field $x2 (ref $P))))
+    (func (export "goWild") (param (ref $R))
+        (struct.set $R $x1
+            (get_local 0)
+            (struct.new $P (i32.const 10) (i32.const 20)))
+    )
+)
+```
+
+will throw when passed a `Point` typed object because of the attempt to
+reinterpret the *nominal* field types of `Rect` with the *structural*
+field types of `$R`.
+
+```js
+WebAssembly.instantiateStreaming(fetch('example3.wasm'))
+.then(({instance}) => {
+    instance.exports.goWild(new Rect(new Point(1,2), new Point(3,4)));  // throws
+});
+```
+
+For the call to succeed, the WebAssembly module must intead *import* the
+exact Type Definition object and use that type import in all relevant
+function signatures:
+
+```wat
+;; example4.wat --> example4.wasm
+(module
+    (type $Point (import "" "Point") (eq (struct (field $x i32) (field $y i32))))
+    (type $Rect (struct (field (ref $Point)) (field (ref $Point))))
+    (func (export "goWild") (param (ref $Rect))
+        (struct.set $R $x1
+            (get_local 0)
+            (struct.new $Point (i32.const 10) (i32.const 20)))
+    )
+)
+```
+
+With this module, the following JS code can instantiate the module with a Type
+Definition object and then later pass it typed object instances of the same Type
+Definition object (but no other):
+
+```js
+function makeTypes() {
+    const Point = new StructType([{name: "x", type: int32}, {name: "y", type: int32}]);
+    const Rect = new StructType([{name: "p1", type: Point.ref}, {name: "p2", type: Point.ref}]);
+    return {Point, Rect};
+}
+
+const {Point, Rect} = makeTypes();
+
+WebAssembly.instantiateStreaming(fetch('example4.wasm'), {'':{Point}})
+.then(({instance}) => {
+    var rect = new Rect(new Point(1,2), new Point(3,4));
+    instance.exports.goWild(rect);  // succeeds
+    assert(rect.x1 instanceof Point);
+    assert(rect.x1.x === 10);
+    assert(rect.x1.y === 20);
+
+    const {Point2:Point, Rect2:Rect} = makeTypes();
+    instance.exports.goWild(new Rect2(new Point2(1,2), new Point2(3,4)));  // throws
```

Comment:
You might want to give a bit more explanation where exactly this would throw -- notably it's the call attempt itself that throws (in the ToWebAssemblyValue coercion), not the struct.set, if I understand correctly.

Reply:
Correct and good point, will do.